Package request: Grayjay Desktop

## Project description

Grayjay is an [..] app that serves as a video player and source aggregator. It allows you to stream and organize videos from various sources, providing a unified platform for your entertainment needs.

https://grayjay.app/#faq

## Metadata

* homepage URL: https://grayjay.app/ and https://grayjay.app/desktop/
* source URL: https://gitlab.futo.org/videostreaming/Grayjay.Desktop
* license: FUTO Source First License 1.1 https://gitlab.futo.org/videostreaming/Grayjay.Desktop/-/blob/master/LICENSE.md
* platforms: linux
---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

Also, it seems that you can't remove the Portable file prior to compilation as it errors out:
```
...
> CSC : error CS2001: Source file '/build/source/Grayjay.Desktop.CEF/Portable' could not be found. [/build/source/Grayjay.Desktop.CEF/Grayjay.Desktop.CEF.csproj]
...
```

Removing it after compiling does work, however.

For completeness, I've attached the updated WIP derivation below, as I noticed I could not build all the subprojects using just the solution (previously the CEF project would not build and it kept looking for CEF in my home directory). They are taken from the [`Grayjay.Desktop.sln` file](https://gitlab.futo.org/videostreaming/Grayjay.Desktop/-/blob/7279b53d81ca62820ba030f34540e6e0ef7d9198/Grayjay.Desktop.sln).
<details>

```nix
{ buildDotnetModule
, fetchFromGitLab
, dotnetCorePackages
}:

buildDotnetModule (finalAttrs: {
  pname = "grayjay";
  version = "3";

  src = fetchFromGitLab {
    domain = "gitlab.futo.org";
    owner = "videostreaming";
    repo = "Grayjay.Desktop";
    rev = "refs/tags/${finalAttrs.version}";
    hash = "sha256-NC6rwgnw0w/87jWU8WyylK/5XMTanWEUDPp9AHdNhI0=";
    fetchSubmodules = true;
  };

  projectFile = [
    "Grayjay.ClientServer/Grayjay.ClientServer.csproj"
    "Grayjay.Engine/Grayjay.Engine/Grayjay.Engine.csproj"
    "Grayjay.Desktop.CEF/Grayjay.Desktop.CEF.csproj"
    "FUTO.MDNS/FUTO.MDNS/FUTO.MDNS.csproj"
    "JustCef/DotCef.csproj"
  ];

  testProjectFile = [
    "Grayjay.Desktop.Tests/Grayjay.Desktop.Tests.csproj"
    "Grayjay.Engine/Grayjay.Engine.Tests/Grayjay.Engine.Tests.csproj"
  ];

  nugetDeps = ./deps.json;

  dotnet-runtime = dotnetCorePackages.aspnetcore_8_0;

  executables = [ "Grayjay" ];

  postInstall = ''
    rm $out/lib/grayjay/Portable
  '';

  # TODO: Add dependencies later once the paths are figured out.
  # runtimeDeps = [];

  meta = {

  };
})
```

</details>

--- a/lib/licenses.nix
+++ b/lib/licenses.nix
@@ -1164,6 +1164,13 @@ lib.mapAttrs mkLicense (
       fullName = "Sendmail License";
     };
 
+    sfl = {
+      fullName = "Source First License 1.1";
+      url = "https://gitlab.futo.org/videostreaming/grayjay/-/blob/master/LICENSE.md";
+      free = false;
+      redistributable = true;
+    };
+
     sgi-b-20 = {
       spdxId = "SGI-B-2.0";
       fullName = "SGI Free Software License B v2.0";

--- a/maintainers/maintainer-list.nix
+++ b/maintainers/maintainer-list.nix
@@ -21209,6 +21209,11 @@
     github = "samemrecebi";
     githubId = 64419750;
   };
+  samfundev = {
+    name = "samfundev";
+    github = "samfundev";
+    githubId = 6759716;
+  };
   samhug = {
     email = "s@m-h.ug";
     github = "samhug";

--- a/pkgs/by-name/gr/grayjay/package.nix
+++ b/pkgs/by-name/gr/grayjay/package.nix
@@ -1,0 +1,191 @@
+{
+  buildDotnetModule,
+  fetchFromGitLab,
+  dotnetCorePackages,
+  buildNpmPackage,
+  lib,
+  libz,
+  icu,
+  openssl,
+  xorg,
+  gtk3,
+  glib,
+  nss,
+  nspr,
+  dbus,
+  atk,
+  cups,
+  libdrm,
+  expat,
+  libxkbcommon,
+  pango,
+  cairo,
+  udev,
+  alsa-lib,
+  mesa,
+  libGL,
+  libsecret,
+  nix-update-script,
+  autoPatchelfHook,
+  makeDesktopItem,
+  copyDesktopItems,
+  libgcc,
+  krb5,
+  wrapGAppsHook3,
+}:
+let
+  version = "5";
+  src = fetchFromGitLab {
+    domain = "gitlab.futo.org";
+    owner = "videostreaming";
+    repo = "Grayjay.Desktop";
+    tag = version;
+    hash = "sha256-xrbYghNymny6MQrvFn++GaI+kUoOphPQMWcqH47U1Yg=";
+    fetchSubmodules = true;
+    fetchLFS = true;
+  };
+  frontend = buildNpmPackage {
+    name = "grayjay-frontend";
+    inherit version src;
+
+    sourceRoot = "source/Grayjay.Desktop.Web";
+
+    npmBuildScript = "build";
+    npmDepsHash = "sha256-pTEbMSAJwTY6ZRriPWfBFnRHSYufSsD0d+hWGz35xFM=";
+
+    installPhase = ''
+      runHook preInstall
+      cp -r dist/ $out
+      runHook postInstall
+    '';
+  };
+in
+buildDotnetModule {
+  pname = "grayjay";
+
+  inherit version src frontend;
+
+  buildInputs = [
+    openssl
+    libgcc
+    xorg.libX11
+    gtk3
+    glib
+    alsa-lib
+    mesa
+    nspr
+    nss
+    icu
+    krb5
+  ];
+
+  nativeBuildInputs = [
+    autoPatchelfHook
+    wrapGAppsHook3
+    copyDesktopItems
+  ];
+
+  dontWrapGApps = true;
+
+  desktopItems = [
+    (makeDesktopItem {
+      name = "Grayjay";
+      exec = "Grayjay";
+      icon = "grayjay";
+      comment = "Cross platform media application for streaming and downloading media";
+      desktopName = "Grayjay Desktop";
+      categories = [ "Network" ];
+    })
+  ];
+
+  projectFile = [
+    "Grayjay.ClientServer/Grayjay.ClientServer.csproj"
+    "Grayjay.Engine/Grayjay.Engine/Grayjay.Engine.csproj"
+    "Grayjay.Desktop.CEF/Grayjay.Desktop.CEF.csproj"
+    "FUTO.MDNS/FUTO.MDNS/FUTO.MDNS.csproj"
+    "JustCef/DotCef.csproj"
+  ];
+
+  testProjectFile = [
+    "Grayjay.Desktop.Tests/Grayjay.Desktop.Tests.csproj"
+    "Grayjay.Engine/Grayjay.Engine.Tests/Grayjay.Engine.Tests.csproj"
+  ];
+
+  nugetDeps = ./deps.json;
+
+  dotnet-runtime = dotnetCorePackages.aspnetcore_8_0;
+
+  executables = [ "Grayjay" ];
+
+  preBuild = ''
+    rm -r Grayjay.ClientServer/wwwroot/web
+    cp -r ${frontend} Grayjay.ClientServer/wwwroot/web
+  '';
+
+  postInstall = ''
+    chmod +x $out/lib/grayjay/cef/dotcefnative
+    chmod +x $out/lib/grayjay/ffmpeg
+    rm $out/lib/grayjay/Portable
+    ln -s /tmp/grayjay-launch $out/lib/grayjay/launch
+    ln -s /tmp/grayjay-cef-launch $out/lib/grayjay/cef/launch
+    mkdir -p $out/share/icons/hicolor/scalable/apps
+    ln -s $out/lib/grayjay/grayjay.png $out/share/icons/hicolor/scalable/apps/grayjay.png
+  '';
+
+  makeWrapperArgs = [
+    "--chdir"
+    "${placeholder "out"}/lib/grayjay"
+  ];
+
+  preFixup = ''
+    makeWrapperArgs+=("''${gappsWrapperArgs[@]}")
+  '';
+
+  runtimeDeps = [
+    libz
+
+    xorg.libXcomposite
+    xorg.libXdamage
+    xorg.libXext
+    xorg.libXfixes
+    xorg.libXrandr
+    xorg.libxcb
+
+    dbus
+    atk
+    cups
+    libdrm
+    expat
+    libxkbcommon
+    pango
+    cairo
+    udev
+    libGL
+    libsecret
+  ];
+
+  passthru.updateScript = nix-update-script {
+    extraArgs = [
+      "--subpackage"
+      "frontend"
+      "--url"
+      "https://github.com/futo-org/Grayjay.Desktop"
+    ];
+  };
+
+  meta = {
+    description = "Cross-platform application to stream and download content from various sources";
+    longDescription = ''
+      Grayjay is a cross-platform application that enables users to
+      stream and download multimedia content from various online sources,
+      most prominently YouTube.
+      It also offers an extensible plugin API to create and import new
+      integrations.
+    '';
+    homepage = "https://grayjay.app/desktop/";
+    license = lib.licenses.sfl;
+    maintainers = with lib.maintainers; [ samfundev ];
+    platforms = [ "x86_64-linux" ];
+    mainProgram = "Grayjay";
+  };
+}

